### PR TITLE
Collections-818: convert to characteristics flag

### PR DESCRIPTION
**Pull request by Claudenw**
converted BloomFilter.isSparse() to BloomFilter.characteristics() and defiend SPARSE constant.

fixes Collections-818

### CODE REVIEW

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java [hunk 1]
```diff
@@ -29,6 +29,15 @@
...
 public interface BloomFilter extends IndexProducer, BitMapProducer {
 
+    /**
+     * The sparse characteristic used to determine the best method for matching.
+     * <p>For `sparse` implementations
+     * the {@code forEachIndex(IntConsumer consumer)} method is more efficient.  For non `sparse` implementations
+     * the {@code forEachBitMap(LongConsumer consumer)} is more efficient.  Implementers should determine if it is easier
+     * for the implementation to produce indexes of bit map blocks.</p>
+     */
+    int SPARSE=0x1;
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java:39]: `SPARSE = 0x1`

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java [hunk 2]
```diff
@@ -38,17 +47,12 @@ public interface BloomFilter extends IndexProducer, BitMapProducer {
...
-     * <p>For `sparse` implementations
-     * the {@code forEachIndex(IntConsumer consumer)} method is more efficient.  For non `sparse` implementations
-     * the {@code forEachBitMap(LongConsumer consumer)} is more efficient.  Implementers should determine if it is easier
-     * for the implementation to produce indexes of bit map blocks.</p>
-     *
-     * @return {@code true} if the implementation is sparse {@code false} otherwise.
-     * @see BitMap
+     * Returns the bitmap of characteristics of the filter.
+     * <p>
+     * Characteristics are defined as bits witin the characteristics integer.
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java:52]: `within`

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java [hunk 3]
```diff
@@ -38,17 +47,12 @@ public interface BloomFilter extends IndexProducer, BitMapProducer {
...
-     * This method is used to determine the best method for matching.
-     *
-     * <p>For `sparse` implementations
-     * the {@code forEachIndex(IntConsumer consumer)} method is more efficient.  For non `sparse` implementations
-     * the {@code forEachBitMap(LongConsumer consumer)} is more efficient.  Implementers should determine if it is easier
-     * for the implementation to produce indexes of bit map blocks.</p>
-     *
-     * @return {@code true} if the implementation is sparse {@code false} otherwise.
-     * @see BitMap
+     * Returns the bitmap of characteristics of the filter.
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java:50]: Drop the `bitmap of`. I think it confusing as `bitmap` in this package is specifically used to refer to a range of indices [0, 63] packed into a long.

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java [hunk 4]
```diff
@@ -69,7 +73,7 @@ public interface BloomFilter extends IndexProducer, BitMapProducer {
      */
     default boolean contains(BloomFilter other) {
         Objects.requireNonNull(other, "other");
-        return isSparse() ? contains((IndexProducer) other) : contains((BitMapProducer) other);
+        return (characteristics()&SPARSE)>0 ? contains((IndexProducer) other) : contains((BitMapProducer) other);
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java:76]: whitespace: `(characteristics() & SPARSE) != 0`

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java [hunk 5]
```diff
@@ -144,7 +148,7 @@ default boolean merge(Hasher hasher) {
         Objects.requireNonNull(hasher, "hasher");
         Shape shape = getShape();
         // create the bloomfilter that is most likely to merge quickly with this one
-        BloomFilter result = isSparse() ? new SparseBloomFilter(shape, hasher) : new SimpleBloomFilter(shape, hasher);
+        BloomFilter result = (characteristics()&SPARSE)>0 ? new SparseBloomFilter(shape, hasher) : new SimpleBloomFilter(shape, hasher);
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java:151]: whitespace: `(characteristics() & SPARSE) != 0`

File: src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java [hunk 6]
```diff
@@ -65,7 +65,7 @@ public SimpleBloomFilter(BloomFilter other) {
         this.shape = other.getShape();
         this.bitMap = new long[BitMap.numberOfBitMaps(shape.getNumberOfBits())];
         this.cardinality = 0;
-        if (other.isSparse()) {
+        if ((other.characteristics()&SPARSE)>0) {
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java:68]: whitespace: `(other.characteristics() & SPARSE) != 0`

File: src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java [hunk 7]
```diff
@@ -194,7 +194,7 @@ public boolean merge(Hasher hasher) {
     @Override
     public boolean merge(BloomFilter other) {
         Objects.requireNonNull(other, "other");
-        if (other.isSparse()) {
+        if ((other.characteristics()&SPARSE)>0) {
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java:197]: whitespace: `(other.characteristics() & SPARSE) != 0`

File: src/main/java/org/apache/commons/collections4/bloomfilter/SparseBloomFilter.java [hunk 8]
```diff
@@ -169,7 +169,7 @@ public boolean merge(Hasher hasher) {
     @Override
     public boolean merge(BloomFilter other) {
         Objects.requireNonNull(other, "other");
-        IndexProducer producer = other.isSparse() ? (IndexProducer) other : IndexProducer.fromBitMapProducer(other);
+        IndexProducer producer = (other.characteristics()&SPARSE)>0 ? (IndexProducer) other : IndexProducer.fromBitMapProducer(other);
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/SparseBloomFilter.java:172]: whitespace: `(other.characteristics() & SPARSE) != 0`

File: src/main/java/org/apache/commons/collections4/bloomfilter/SparseBloomFilter.java [hunk 9]
```diff
@@ -58,7 +58,7 @@ public SparseBloomFilter(BloomFilter other) {
         Objects.requireNonNull(other, "other");
         this.shape = other.getShape();
         this.indices = new TreeSet<>();
-        if (other.isSparse()) {
+        if ((other.characteristics()&SPARSE)>0) {
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/SparseBloomFilter.java:61]: whitespace: `(other.characteristics() & SPARSE) != 0`

**Review by aherbert — CHANGES_REQUESTED** [review]
Changes look good apart from some minor issues.

Note that it is more conventional to compare the bit operation is not zero to detect if a flag is set. This is how it is done for example in the Bloom filter `BitMap` utility class in the `contains` method so we should be consistent.

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java (hunk 1 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java (hunk 2 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java (hunk 4 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java (hunk 6 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/SparseBloomFilter.java (hunk 8 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/SparseBloomFilter.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java (hunk 3 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java (hunk 5 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/BloomFilter.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java (hunk 7 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/SparseBloomFilter.java (hunk 9 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/SparseBloomFilter.java]: fixed

**Pull request merged**